src/hash.h

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright (c) 2019, Intel Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <vector>
+
+#include "enums.h"
+
+namespace yarpgen {
+
+const size_t HASH_SEED = 42;
+
+// Class that is used to calculate various hashes.
+class Hash {
```

Comment:
I'm not following how Hash assumed to be used and for what types.
Is it only to use in folding set? Why do we need it then? Why defaulf hash is not good enough?

Is it going to be used to compute hash of the IR values in the program?

Current implementation is not covering all types. For not fundamental types you can deal only with with enums in current implementation. Is it by design?

Can we reinterpret any type as a sequence of bytes and hash them?

Reply:
Right now it is used only in folding sets. I don't think we want to use it for other IR types, because the majority of them will be encapsulated in std::shared_ptr. We can use a memory address as a unique id or hash.

It simplifies code and the reasoning is the following:
1. [std::hash](https://en.cppreference.com/w/cpp/utility/hash) requires to specify a type (e.g. std::hash<int>()(a)). With templates, you don't need to do this.
2. It is simpler to create an object of type Hash and pass the object that we want to hash to it. We don't need to call hash, hashCombine and pass seed each time.

Therefore it basically allows you to create an instance and pass all of the objects that you want to hash without thinking about anything. 

C++ types that we store in IR Types are fundamental types, enums, and vectors. We don't need to carry about anything else.

We can reinterpret any type as a sequence of bytes, but I think it is an overkill for such a simple task.

Reply:
I'm not against templates, but it was not clear to me if you were going to work with outher types. If you are going to revert the code to templates, make sure that you have a static_assert() that the type is supported.

Also add a comment that this class is intended to support folding sets, just to be clear about it's purpose.

Reply:
Done